dhis-2/dhis-services/dhis-service-administration/src/main/resources/data-integrity-checks/programs/program_stages_no_programs.yaml

```diff
@@ -0,0 +1,74 @@
+# Copyright (c) 2004-2022, University of Oslo
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+# Redistributions of source code must retain the above copyright notice, this
+# list of conditions and the following disclaimer.
+#
+# Redistributions in binary form must reproduce the above copyright notice,
+# this list of conditions and the following disclaimer in the documentation
+# and/or other materials provided with the distribution.
+# Neither the name of the HISP project nor the names of its contributors may
+# be used to endorse or promote products derived from this software without
+# specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+---
+name: program_stages_no_programs
+description: Program stages with no associated program
+section: Programs
+section_order: 2
+summary_sql: >-
+  WITH rs as (
+  SELECT a.uid,a.name, b.name as comment
+  from programstage a
+    INNER JOIN program b ON a.programid = b.programid
+  WHERE a.programid IS NULL
+  )
+  select count(*) as value,
+  100.0 * count(*) / NULLIF( (select count(*) from programstage),0) as percent
+  from rs;
+details_sql: >-
+  SELECT a.uid,a.name, b.name as comment
+  from programstage a
+    INNER JOIN program b ON a.programid = b.programid
+  WHERE a.programid IS NULL;
+severity: SEVERE
+introduction: >
+  All program stages should be associated with a program. This check identifies
+  program stages that are not linked to any program. Certain version of DHIS2
+  do now allow program stages to exist without being linked to a program and may
+  result in the system not being able to start.
+details_id_type: programStages
+recommendation: >
+  You will need to either delete the program stage or link it to a program.
```

Comment:
Here in recommendation I would suggest to check if there is data attached to the faulty program stages (provide a simple query to check that) and if it is the case, suggest to remove the program stage.
If there is any data attached, then leave it to the user if they want to try to find a program or delete all the data and then the program stage.

Reply:
I added some details in the "comments" section of the details query which indicates whether data exists or not. At the moment, it is not possible to see the "comments" in the user interface though.